File: scripts/tech-ds/turret/electric-storm-turret.js
// Copyright (C) 2020 abomb4
//
// This file is part of Dimension Shard.
//
// Dimension Shard is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Dimension Shard is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Dimension Shard.  If not, see <http://www.gnu.org/licenses/>.

const lib = require('abomb4/lib');
const blockTypes = require('abomb4/block-types');

const items = require('ds-common/items');
const {
    dimensionAlloy
} = items;
const dsGlobal = require('ds-common/ds-global');

const {
    newElectricStormBulletType
} = require('ds-common/bullet-types/index');
var i = 1;
const turret = blockTypes.newNoRotatingTurret({
    name: 'electric-storm-turret',
    turretType: ItemTurret,
    buildVisibility: BuildVisibility.shown,
    category: Category.turret,
    liquidCapacity: 120,
    health: 5200,
    size: 5,
    reloadTime: 150 - 1,
    range: 384,
    inaccuracy: 6,
    spread: 45,
    shots: 8,
    rotateSpeed: 1000,
    velocityInaccuracy: 0.8,
    shootCone: 360,
    heatColor: Color.valueOf('69dcee'),
    shootSound: lib.loadSound('electric-shot'),
    requirements: ItemStack.with(
        Items.copper, 2500,
        Items.lead, 2000,
        Items.metaglass, 1500,
        Items.silicon, 1200,
        Items.plastanium, 1000,
        Items.phaseFabric, 800,
        Items.surgeAlloy, 1000,
        items.dimensionShard, 1600,
        items.hardThoriumAlloy, 800,
        items.dimensionAlloy, 300
    ),
    coolantMultiplier: 0.15,
    coolantUsage: 2,
    blockOverrides: {
        middleRegion: lib.loadRegion('electric-storm-turret-middle'),
        middleHeatRegion: lib.loadRegion('electric-storm-turret-middle-heat'),
        load() {
            this.super$load();
            this.baseRegion = lib.loadRegion('electric-storm-turret-base');
            this.middleRegion = lib.loadRegion('electric-storm-turret-middle');
            this.middleHeatRegion = lib.loadRegion('electric-storm-turret-middle-heat');
        },
        isPlaceable() { return dsGlobal.techDsAvailable() && this.super$isPlaceable(); },
        drawPlace(x, y, rotation, valid) {
            if (!dsGlobal.techDsAvailable()) {
                this.drawPlaceText(lib.getMessage("msg", "dimensionCoreRequired"), x, y, valid);
            }
            this.super$drawPlace(x, y, rotation, valid);
        },
    },
    buildingOverrides: () => ({
        pigu: i++,
        drawRotate: 0,
        drawRotateAccel: 0,
        // I think the default udpatShooting and updateCooling is wrong, so modify it.
        updateShooting() {
            if (this.reload >= this.block.reloadTime) {
                var type = this.peekAmmo();
                this.shoot(type);
                this.reload -= this.block.reloadTime;
            }
        },
        updateTile() {
            this.super$updateTile();
            // Do reload if has ammo.
            if (this.hasAmmo() && this.reload < this.block.reloadTime) {
                this.reload += this.delta() * this.peekAmmo().reloadMultiplier * this.baseReloadSpeed();
            }
            if (this.hasAmmo() && this.cons.valid()) {
                this.drawRotateAccel = Mathf.lerpDelta(this.drawRotateAccel, 0.4, 0.02);
            } else {
                this.drawRotateAccel = Mathf.lerpDelta(this.drawRotateAccel, 0, 0.02);
            }
            this.drawRotate += this.drawRotateAccel
        },
        bullet(type, angle) {
            const { x, y, targetPos, team } = this;
            const { tr, velocityInaccuracy, minRange } = this.block;
            var lifeScl = type.scaleVelocity ? Mathf.clamp(Mathf.dst(x + tr.x, y + tr.y, targetPos.x, targetPos.y) / type.range(), minRange / type.range(), range / type.range()) : 1;
            // type.create(this, team, x + tr.x, y + tr.y, angle, -1, 1 + Mathf.range(velocityInaccuracy), lifeScl, {
            //     target: !this.isControlled() && this.target ? this.target : new Vec2(targetPos.getX(), targetPos.getY())
            // });

            if ((targetPos && targetPos.getX() != 0 && targetPos.getY() != 0) && (this.isControlled() || this.logicShooting)) {
                type.create(this, team, x + tr.x, y + tr.y, angle, -1, 1 + Mathf.range(velocityInaccuracy), lifeScl, {
                    target: targetPos
                    // target: new Vec2(targetPos.getX(), targetPos.getY())
                });
            } else {
                type.create(this, team, x + tr.x, y + tr.y, angle, 1 + Mathf.range(velocityInaccuracy), lifeScl);
            }
        },
        draw() {
            const {
                baseRegion, region, size, heatColor, heatRegion, middleRegion, middleHeatRegion
            } = turret;
            const { x, y, drawRotate } = this;
            Draw.rect(baseRegion, x, y);
            Draw.color();

            Draw.rect(middleRegion, this.x, this.y);
            if (this.heat > 0.00001) {
                Draw.color(heatColor, this.heat);
                Draw.blend(Blending.additive);
                Draw.rect(middleHeatRegion, this.x, this.y);
                Draw.blend();
                Draw.color();
            }

            Draw.z(Layer.turret);
            Draw.rect(region, this.x, this.y, drawRotate - 90);

            if (this.heat > 0.00001) {
                Draw.color(heatColor, this.heat);
                Draw.blend(Blending.additive);
                Draw.rect(heatRegion, this.x, this.y, drawRotate - 90);
                Draw.blend();
                Draw.color();
            }
        },
    })
});

turret.ammo(dimensionAlloy, newElectricStormBulletType({
    speedStart: 0.8,
    homingDelay: 30,
    lifetime: 150,
    homingRange: 400,
    ammoMultiplier: 2,
}));

turret.consumes.powerCond(75, boolf(b => b.isActive()));

exports.electricStormTurret = turret;

File: src/dimensionshard/world/blocks/ElectricStormTurret.java
package dimensionshard.world.blocks;

import arc.graphics.Color;
import arc.graphics.g2d.TextureRegion;
import arc.math.Mathf;
import dimensionshard.DsGlobal;
import dimensionshard.DsItems;
import dimensionshard.entities.bullets.ElectricStormBulletType;
import dimensionshard.libs.Lib;
import mindustry.content.Items;
import mindustry.entities.Mover;
import mindustry.entities.bullet.BulletType;
import mindustry.entities.pattern.ShootSpread;
import mindustry.gen.Bullet;
import mindustry.type.Category;
import mindustry.type.ItemStack;
import mindustry.world.blocks.defense.turrets.ItemTurret;
import mindustry.world.meta.BuildVisibility;

/**
 * 闪电风暴炮塔
 *
 * @author abomb4 2022-10-08
 */
public class ElectricStormTurret extends ItemTurret {

    /** 中心 */
    public TextureRegion middleRegion;
    /** 中心发热 */
    public TextureRegion middleHeatRegion;
    /** 垂直方向速度随机 */
    public float velocityInaccuracy = 0.8F;

    public ElectricStormTurret(String name) {
        super(name);
        buildVisibility = BuildVisibility.shown;
        category = Category.turret;
        liquidCapacity = 120;
        health = 7200;
        size = 5;
        reload = 150 - 1;
        range = 400;
        inaccuracy = 6;
        this.shoot = new ShootSpread(8, 45);
        rotateSpeed = 1000;
        // velocityInaccuracy = 0.8;
        shootCone = 360;
        heatColor = Color.valueOf("69dcee");
        shootSound = Lib.loadSound("electric-shot");
        requirements = ItemStack.with(
            Items.copper, 2500,
            Items.lead, 2000,
            Items.metaglass, 1500,
            Items.silicon, 1200,
            Items.plastanium, 1000,
            Items.phaseFabric, 800,
            Items.surgeAlloy, 1000,
            DsItems.dimensionShard, 1600,
            DsItems.hardThoriumAlloy, 800,
            DsItems.dimensionAlloy, 300
        );
        coolantMultiplier = 0.15F;
        consumeCoolant(2);
        ammo(DsItems.dimensionAlloy, new ElectricStormBulletType() {{
            speedStart = 0.8F;
            homingDelay = 30;
            lifetime = 150;
            homingRange = 4000;
            ammoMultiplier = 2;
        }});
        consumePowerDynamic(building -> ((ElectricStormTurretBuild) building).isActive() ? 75 : 0);
    }

    @Override
    public void load() {
        super.load();
        this.middleRegion = Lib.loadRegion("electric-storm-turret-middle");
        this.middleHeatRegion = Lib.loadRegion("electric-storm-turret-middle-heat");
    }

    @Override
    public boolean isPlaceable() {
        return DsGlobal.techDsAvailable() && super.isPlaceable();
    }

    @Override
    public void drawPlace(int x, int y, int rotation, boolean valid) {
        if (!DsGlobal.techDsAvailable()) {
            this.drawPlaceText(Lib.getMessage("msg", "dimensionCoreRequired"), x, y, valid);
            return;
        }
        super.drawPlace(x, y, rotation, valid);
    }

    public class ElectricStormTurretBuild extends ItemTurret.ItemTurretBuild {
        /** 旋转角度 */
        public float drawRotate = 0;
        /** 加速度 */
        public float drawRotateAccel = 0;

        @Override
        public void updateTile() {
            super.updateTile();
            // rotate if has ammo
            if (this.hasAmmo()) {
                this.drawRotateAccel = Mathf.lerpDelta(this.drawRotateAccel, 0.4F, 0.02F);
            } else {
                this.drawRotateAccel = Mathf.lerpDelta(this.drawRotateAccel, 0, 0.02F);
            }
            this.drawRotate += this.drawRotateAccel;
        }

        @Override
        protected void bullet(BulletType type, float xOffset, float yOffset, float angleOffset, Mover mover) {
            super.bullet(type, xOffset, yOffset, angleOffset, mover);
        }

        @Override
        protected void handleBullet(Bullet bullet, float offsetX, float offsetY, float angleOffset) {
            if ((targetPos != null && targetPos.getX() != 0 && targetPos.getY() != 0)
                && (this.isControlled() || this.logicShooting)
                && bullet.type instanceof ElectricStormBulletType ebullet) {
                ((ElectricStormBulletType.ElectricStormBulletData) bullet.data).target = targetPos;
            }
        }
    }
}
